src/inc/til/latch.h

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+#ifdef __cpp_lib_latch
+#include <latch>
+#endif
+
+namespace til
+{
+#ifdef __cpp_lib_latch
+    using latch = std::latch;
+#else
+    class latch
+    {
+    public:
+        [[nodiscard]] static constexpr ptrdiff_t max() noexcept
+        {
+            return std::numeric_limits<ptrdiff_t>::max();
+        }
+
+        constexpr explicit latch(const ptrdiff_t expected) noexcept :
+            counter{ expected }
+        {
+            assert(expected >= 0);
+        }
+
+        latch(const latch&) = delete;
+        latch& operator=(const latch&) = delete;
+
+        void count_down(const ptrdiff_t n = 1) noexcept
+        {
+            assert(n >= 0);
+
+            const auto old = counter.fetch_sub(n, std::memory_order_release);
+            if (old == n)
+            {
+                WakeByAddressAll((PVOID)&counter);
+                return;
+            }
+
+            assert(old > n);
+        }
+
+        [[nodiscard]] bool try_wait() const noexcept
+        {
+            return counter.load(std::memory_order_acquire) == 0;
+        }
+
+        void wait() const noexcept
+        {
+            while (true)
+            {
+                auto current = counter.load(std::memory_order_acquire);
+                if (current == 0)
+                {
+                    return;
+                }
+
+                assert(current > 0);
+                WaitOnAddress((PVOID)&counter, &current, sizeof(counter), INFINITE);
```

Comment:
nit: we usually avoid C-style casts; whether you want to fix this is up to you

Reply:
It seems that WaitOnAddress itself should be declared as taking `const volatile VOID *Address`, but isn't. Perhaps such a change in the Windows SDK would be too breaking but Microsoft now has an opportunity to fix that in win32metadata, similar to what they did in <https://github.com/microsoft/win32metadata/issues/449>.

Reply:
I've used this function many times in the past and this is actually something that has always bothered me. I'll try to submit a change request for this function!
But it'll take some time, as I'm still unfamiliar with working on the internal OS repository and I'm not sure if such a signature change would be accepted in the first place.